src/python/pants/option/global_options.py

```diff
@@ -453,6 +458,8 @@ def from_options(cls, options: OptionValueContainer) -> LocalStoreOptions:
     remote_instance_name=None,
     remote_ca_certs_path=None,
     # Process execution setup.
+    child_process_max_memory_usage=memory_size("1GiB"),
+    child_process_default_memory_usage=memory_size("256MiB"),
```

Comment:
Defaults are really interesting here: I wonder if we should perhaps _not_ cap `max_memory_usage` by default, and instead continue to spawn based on parallelism by default (so on an 8 core machine, you would get `16` processes, each using 256MB)?

Because the effect of setting max_memory_usage by default is that machine differences _require_ you to configure on a per-machine basis to roughly match your parallelism. And while capping memory usage will be useful/necessary in some environments, requiring it by default might be a pain.

Reply:
it's a good point, in fact these defaults have been chosen totally arbitrarily. In fact while dealing with this I have been wondering whether I'm misinterpretting the meaning of these two settings.

I'm interpretting the `max_memory_usage` as a total cap for all potential processes spawned, instead of the max memory per-process. A max cap for all processes has potential when running pants in constrained environments and implementation-wise would translate passing the flags `-Xmx{calculated_value}mb` and `-Xms{default_size}mb` to each nailgun instance such that they all can fit inside in the total cap given in the options.

So for the cases in which there is a max cap value given, then the pool size would be calculated based on the parallelism and would not pass any `-Xmx...` flag to the JVM.

Does this sound reasonable?

Reply:
> I'm interpretting the `max_memory_usage` as a total cap for all potential processes spawned, instead of the max memory per-process.

That's correct.

> A max cap for all processes has potential when running pants in constrained environments and implementation-wise would mean calculating a max-memory per process factoring in parallelism and passing the flags `-Xmx{calculated_value}mb` and `-Xms{default_size}mb` to each nailgun instance such that they all can fit inside in the total cap given in the options.

So, there is one very important bit missing here, which is that: it's important for reproducibility to use a consistent max per-process heap size across all consumers of a repository. Otherwise, compilation and testing might fail on one machine, and succeed on another for a particular target. So the `child_process_default_memory_usage` value needs to be constant/static, and cannot be calculated.

> So for the cases in which there is not a max cap value given, then the pool size would be calculated based only on the parallelism and would not pass any `-Xmx...` flag to the JVM.

The first part is correct, but as to `-Xmx`: we should always pass the `child_process_default_memory_usage` as a templated `-Xmx` value, for the reason given above (reproducibility).

Reply:
(so "_default" is a misnomer: should replace with "_max" probably)

Reply:
ah right, got it now, that's where my misunderstanding was, as I was interpretting `child_process_default_memory_usage` as the value to pass as the `-Xms...` flag, not as a per-process cap. Now it makes more sense to me. Thanks!